pretext/cli.py

```diff
@@ -600,71 +554,82 @@ def generate(
 )
+@click.option(
+    "-s",
+    "--stop-server",
+    is_flag=True,
+    default=False,
+    help="Stop the local http server if running.",
+)
 def view(
     target_name: str,
     access: Literal["public", "private"],
     port: Optional[int],
-    directory: Optional[str],
-    watch: bool,
     build: bool,
     generate: Optional[str],
     no_launch: bool,
+    stop_server: bool,
 ) -> None:
     """
     Starts a local server to preview built PreTeXt documents in your browser.
     TARGET is the name of the <target/> defined in `project.ptx`.
+    If a server is already running, no new server will be started.
     """
-    if directory is not None:
-        if utils.cocalc_project_id() is not None:
-            try:
-                subdir = Path(directory).relative_to(Path.home())
-            except ValueError:
-                subdir = Path()
-            log.info("Directory can be previewed at the following link at any time:")
-            log.info(f"    https://cocalc.com/{utils.cocalc_project_id()}/raw/{subdir}")
-            return
-        port = port or 8000
-        utils.run_server(Path(directory), access, port, no_launch=no_launch)
-        return
     if utils.no_project(task="view the output for"):
         return
     project = Project.parse()
-    target = project.get_target(name=target_name)
-    if target is None:
+    try:
+        target = project.get_target(name=target_name)
+    except AssertionError as e:
         utils.show_target_hints(target_name, project, task="view")
         log.critical("Exiting.")
+        log.debug(e, exc_info=True)
         return
-    # Easter egg to spin up a local server at a specified directory:
-    if utils.cocalc_project_id() is not None:
+
+    port = port or 8128
+    # Call generate if flag is set
+    if generate:
+        try:
+            target.generate_assets(only_changed=False)
+        except Exception as e:
+            log.info(f"Failed to generate assets: {e}")
+            log.debug("", exc_info=True)
+    if build:
         try:
-            subdir = target.output_dir_abspath().relative_to(Path.home())
-        except ValueError:
-            subdir = Path()
-        log.info("Built project can be previewed at the following link at any time:")
-        log.info(f"    https://cocalc.com/{utils.cocalc_project_id()}/raw/{subdir}")
+            target.build()
+        except Exception as e:
+            log.info(f"Failed to build: {e}")
+            log.debug("Exception info:\n##################\n", exc_info=True)
+    # Start server if there isn't one running already:
+    if not utils.server_running():
+        log.info("Starting server.")
+        try:
+            server = project.server_process(
+                output_dir=target.output_dir,
+                access=access,
+                port=port,
+                launch=not no_launch,
+            )
+            server.start()
+        except KeyboardInterrupt:
```

Comment:
Shouldn't this be something like

```python
try:
    while True:
        sleep(1)
except KeyboardInterrupt:
    server.stop()
```

Reply:
Maybe instead of `while True`:

```python
while server.is_alive():
````

Reply:
Yes!  I think I understand what is happening better now too.  This is fixed.